src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java

```diff
@@ -20,6 +21,47 @@ public ConnectedComponentsCalculator(Tangle tangle, int maxTransactions) {
         this.random = new SecureRandom();
     }
 
+    public Collection<Set<Hash>> getConnectedComponents(Set<Hash> transactions) throws Exception {
+        Set<Hash> unvisited = new HashSet<>(transactions);
```

Comment:
An alternative would be using `DAGHelper::traverseApprovees` (and thus BFS) to compute connected components and avoid code duplication.

Reply:
The reason I didn't do that was that I want a bidirectional scan (towards both approvers and approvees), and DAGHelper doesn't provide that. I don't know whether the transaction I'm starting from in each component is a tip or somewhere in the middle.